grift/loaders.py

```diff
@@ -67,6 +69,96 @@ def reload(self):
         self._source = os.environ
 
 
+class VaultException(Exception):
+    """Custom exception raised if Vault response contains error message"""
+
+
+class VaultTokenLoader(DictLoader):
+    def __init__(self, url, path, token):
+        """Load secrets from a Vault path, using token authentication
+
+        See https://www.vaultproject.io/docs
+
+        Args:
+            url: Vault url
+            path: Vault path to fetch secrets from
+            vault_token: token (must have access to vault path)
+        """
+        self._vault_url = url
+        self._path = path
+        self._token = token
+        source_dict = self._fetch_secrets()
+        super(VaultTokenLoader, self).__init__(source_dict)
+
+    @property
+    def _headers(self):
+        """Return token header to access vault"""
+        return {'X-Vault-Token': self._token}
+
+    def _fetch_secrets(self):
+        """Read data from the vault path"""
+        url = '{}/v1/{}'.format(self._vault_url, self._path)
+        resp = requests.get(url, headers=self._headers)
+        resp.raise_for_status()
+        data = resp.json()
+        if data.get('errors'):
+            raise VaultException(u'Error fetching Vault secrets from path {}: {}'
+                                 .format(self._path, data['errors']))
+        return data['data']
+
+    def reload(self):
+        """Reread secrets from the vault path"""
+        self._source = self._fetch_secrets()
+
+    def lookup_token(self):
+        """Convenience method: look up the vault token"""
+        url = '{}/v1/auth/token/lookup-self'.format(self._vault_url)
+        resp = requests.get(url, headers=self._headers)
+        resp.raise_for_status()
+        data = resp.json()
+        if data.get('errors'):
+            raise VaultException(u'Error looking up Vault token: {}'.format(data['errors']))
+        return data
+
+    def renew_token(self):
+        """Convenience method: renew Vault token"""
+        url = '{}/v1/auth/token/renew-self'.format(self._vault_url)
+        resp = requests.get(url, headers=self._headers)
+        resp.raise_for_status()
+        data = resp.json()
+        if data.get('errors'):
+            raise VaultException(u'Error renewing Vault token: {}'.format(data['errors']))
+        return data
+
+
+class VaultAppRoleLoader(VaultTokenLoader):
+    def __init__(self, url, path, role_id, secret_id):
+        """Load secrets from a Vault path, using the AppRole auth backend
+
+        See https://www.vaultproject.io/docs/auth/approle.html
+
+        Args:
+            url: Vault url
+            path: Vault path to fetch secrets from
+            role_id: Vault RoleID
+            secret_id: Vault SecretID
+        """
+        self._role_id = role_id
+        self._secret_id = secret_id
+        token = self._fetch_token()
```

Comment:
Same comment as above, on calling methods before the object is fully constructed.